pkg/query-service/app/queryBuilder/query_builder.go

```diff
@@ -152,11 +153,28 @@ func (qb *QueryBuilder) PrepareQueries(params *v3.QueryRangeParamsV3, args ...in
 					}
 					queries[queryName] = queryString
 				case v3.DataSourceLogs:
-					queryString, err := qb.options.BuildLogQuery(params.Start, params.End, compositeQuery.QueryType, compositeQuery.PanelType, query)
-					if err != nil {
-						return nil, err
+					// if panel type is timeseries/ table, handle limit differently using two queries
+					if (compositeQuery.PanelType == v3.PanelTypeGraph || compositeQuery.PanelType == v3.PanelTypeTable) && query.Limit > 0 && len(query.GroupBy) > 0 {
+						// return two queries
+						limitQuery, err := qb.options.BuildLogQuery(params.Start, params.End, compositeQuery.QueryType, compositeQuery.PanelType, query, constants.FirstQueryGraphLimit)
+						if err != nil {
+							return nil, err
+						}
+
+						placeholderQuery, err := qb.options.BuildLogQuery(params.Start, params.End, compositeQuery.QueryType, compositeQuery.PanelType, query, constants.SecondQueryGraphLimit)
+						if err != nil {
+							return nil, err
+						}
+						queries[queryName+constants.FirstQuerySuffix] = limitQuery
+						queries[queryName+constants.SecondQuerySuffix] = placeholderQuery
```

Comment:
Since you are using the suffixes, will it work for the formula such as `A/2` or similar? The formula will look for `A` whereas the `queries` map contains `A-#first` and `A-#second`?

Reply:
Ahh it won't work as of now, thinking of ways we can solve this.

Any suggestions on ways to implement this as we need the first query to be complete so that that the second query can be formed which can be used in the expression evaluation.

Reply:
Can you run all of the pre queries first and save the result and then prepare the final queries to be executed?